src/server/tools/__init__.py

```diff
@@ -0,0 +1,301 @@
+from os import stat
+from random import choice
+from shutil import move
+from string import ascii_lowercase
+from time import time
+from urllib.parse import unquote_plus
+
+# Third party library imports
+from psycopg2 import connect, OperationalError, ProgrammingError
+from requests import Session
+from requests.exceptions import ConnectionError
+from web import ctx, header
+
+# Own library
+from ssh_utils import Authority
+
+def get_principals(sql_result, username, shell=False):
+    """
+    Transform sql principals into readable one
+    """
+    if sql_result is None or sql_result == '':
+        if shell:
+            return username
+        return [username]
+    else:
+        if shell:
+            return sql_result
+        return sql_result.split(',')
+
+def pretty_ssh_key_hash(pubkey_fingerprint):
+    """
+    Returns a pretty json from raw pubkey
+    KEY_BITS KEY_HASH [JERK] (AUTH_TYPE)
+    """
+    try:
+        key_bits = int(pubkey_fingerprint.split(' ')[0])
+    except ValueError:
+        key_bits = 0
+    except IndexError:
+        key_bits = 0
+
+    try:
+        key_hash = pubkey_fingerprint.split(' ')[1]
+    except IndexError:
+        key_hash = pubkey_fingerprint
+
+    try:
+        auth_type = pubkey_fingerprint.split('(')[-1].split(')')[0]
+    except IndexError:
+        auth_type = 'Unknown'
+
+    rate = 'UNKNOWN'
+    if auth_type == 'DSA':
+        rate = 'VERY LOW'
+    elif (auth_type == 'RSA' and key_bits >= 4096) or (auth_type == 'ECDSA' and key_bits >= 256):
+        rate = 'HIGH'
+    elif auth_type == 'RSA' and key_bits >= 2048:
+        rate = 'MEDIUM'
+    elif auth_type == 'RSA' and key_bits < 2048:
+        rate = 'LOW'
+    elif auth_type == 'ED25519' and key_bits >= 256:
+        rate = 'VERY HIGH'
+
+    return {'bits': key_bits, 'hash': key_hash, 'auth_type': auth_type, 'rate': rate}
+
+def random_string(string_length=10):
+    """Generate a random string of fixed length """
+    letters = ascii_lowercase
+    return ''.join(choice(letters) for i in range(string_length))
+
+def response_render(message, http_code='200 OK', content_type='text/plain'):
+    """
+    This function returns a well-formed HTTP response
+    """
+    header('Content-Type', content_type)
+    ctx.status = http_code
+    return message
+
+def str2date(string):
+    """
+    change xd => seconds
+    change xh => seconds
+    """
+    delta = 0
+    if 'd' in string:
+        delta = timedelta(days=int(string.split('d')[0])).total_seconds()
+    elif 'h' in string:
+        delta = timedelta(hours=int(string.split('h')[0])).total_seconds()
+    return delta
+
+def unquote_custom(string):
+    """
+    Returns True is the string is quoted
+    """
+    if ' ' not in string:
+        string = unquote_plus(string)
+        if '+' in string and '%20' in string:
+            # Old custom quote
+            string = string.replace('%20', ' ')
+    return string
+
+
+class Tools(object):
+    """
+    Class tools
```

Comment:
Tools seems a bit too blurry to me... It does a lot of things and the description is not clear enough.

### 1st 

You did not include these 2 functions which are still in the main python file ?
- `ldap_authentification`
- `data2map`


### 2nd, about the scopes

What do you think about:
* A `Server` Class that has the CASSH specific methods
  - `list_keys`
  - `sign_keys`
* A `ClusterController` Class with:
  - `status` (`cluster_alived`)
  - `get_krl` (`cluster_last_krl`)
  - `update_krl` (`cluster_update_krl`)

Maybe `Server` would merge into `ClusterController` at some point and just be renamed as `Controller` with a option / param to support cluster mode.

And finally the `Tools` class could be removed and these methods would go back as functions :
  - `pg_connection`
  - `get`
  - `post`
  - `sql_to_json`

and `tools`  might me renamed as `cassh_utils` for the functions needed outside of him ?

Reply:
I's a good idea. I'm creating an issue to clean the code in the future, with these insight

Reply:
https://github.com/nbeguier/cassh/issues/71